### PR TITLE
HLRC: Add ILM Status to HLRC

### PR DESCRIPTION
**Pull request by jakelandis**
Adds support for the Index Lifecycle Management Status to the Java
High-Level Rest Client.

Relates to #33100

--------
Notes to reviewer(s):
* I don't think there should be anything controversial here, and probably don't need 5 reviewers ... just wasn't sure who to tag. 
* the naming conventions were chosen to match the start/stopILM.
* Using a TimedRequest instead of a StatusILMRequest... not sure the preference here, but it kinda matches the `AcknowledgedResponse` in terms of not creating typed request/response when not needed.

### CODE REVIEW

File: client/rest-high-level/src/main/java/org/elasticsearch/client/IndexLifecycleClient.java
Comment (gwbrown) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/IndexLifecycleClient.java]: Method names should start with a lowercase letter, so this should be `statusILM`.

Or maybe `getILMStatus`? `statusILM` doesn't make much sense seeing it for the first time - there's no verb there.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/IndexLifecycleClient.java
Comment (gwbrown) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/IndexLifecycleClient.java]: As above re: method names.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/IndexLifecycleClient.java
Comment (dakrone) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/IndexLifecycleClient.java]: I think these should be `LifecycleManagementStatusResponse`, rather than using the ILM in method/class names

Also, the method name is `StatusILM` and we don't capitalize any of our method names

File: client/rest-high-level/src/main/java/org/elasticsearch/client/IndexLifecycleClient.java
Comment (dakrone) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/IndexLifecycleClient.java]: Same here about naming and capitalization

File: client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/StatusILMResponse.java
Comment (dakrone) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/StatusILMResponse.java]: I don't think we need these generated comments

File: client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/StatusILMResponse.java
Comment (gwbrown) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/StatusILMResponse.java]: #33250 moves the existing `OperationMode` class into `o.e.client.indexlifecycle` - you may want to wait until that gets merged (should be later today) and use that one instead of having this as an inner class.

**Review by dakrone — CHANGES_REQUESTED** [review]
I left comments renaming these methods/classes, I'm in favor of:

- LifecycleManagementStatusResponse
- `lifecycleManagementStatus(...)`

File: client/rest-high-level/src/test/java/org/elasticsearch/client/IndexLifecycleIT.java
Comment (gwbrown) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/IndexLifecycleIT.java]: While I get the logic behind why you did it this way, I think generic classes make more sense for responses than requests.  As a user of the HLRC, this feels really strange - even if `GetILMStatusRequest` would be an extra class that doesn't really add anything, I think having separate classes for each request type is the way to go.  I could be wrong though, I'd like to hear others' input on this.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/indexlifecycle/StatusILMResponseTests.java
Comment (gwbrown) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/indexlifecycle/StatusILMResponseTests.java]: Using the shared class would let you get rid of this test, too.

**Review by AthenaEryma — CHANGES_REQUESTED** [review]
Left some comments.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/IndexLifecycleClient.java
Comment (jakelandis) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/IndexLifecycleClient.java]: oops ... IDE find and replace gone wild. will fix.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/IndexLifecycleClient.java
Comment (jakelandis) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/IndexLifecycleClient.java]: > Also, the method name is StatusILM and we don't capitalize any of our method names

oops didn't meant to capitalize the method name. (kinda surprised that check style didn't catch that). will fix. 

> I think these should be LifecycleManagementStatusResponse, rather than using the ILM in method/class names

I tend to agree, but was matching the start/stopILM here:  https://github.com/elastic/elasticsearch/tree/index-lifecycle/x-pack/protocol/src/main/java/org/elasticsearch/protocol/xpack/indexlifecycle 

Should I rename those too ?

File: client/rest-high-level/src/test/java/org/elasticsearch/client/IndexLifecycleIT.java
Comment (jakelandis) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/IndexLifecycleIT.java]: I don't have a strong opinion beyond consistency in the API.  

Perhaps we should allow a no-arg method in the client that creates the request for them. However, that also creates inconsistency (unless you create a bunch of new methods).

> I'd like to hear others' input on this.

..same

File: client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/StatusILMResponse.java
Comment (jakelandis) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/StatusILMResponse.java]: thanks! ... will wait till that gets merged and update accordingly.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/indexlifecycle/StatusILMResponseTests.java
Comment (jakelandis) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/indexlifecycle/StatusILMResponseTests.java]: will wait for #33250 to use shared class.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/StatusILMResponse.java
Comment (jakelandis) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/StatusILMResponse.java]: I can remove them.  FWIW, I generally put these in here as a way to signal nothing special here, so no tests, and if changes are made need to re-generate.

**Comment by jakelandis** [comment]
Changes pushed:
* Fixed incorrect method uppercase 
* Renamed statsILM -> lifecycleManagementStatus
* removed "generated" comments

TODO (this PR):
* Integrate with #33250 once merged
* Decision on named Request classes vs. generic TimedRequest class for empty requests [1]

TODO (not in this PR):
* rename stopILM and startILM ? [2]
* enforce basic method naming via checkstyle

@hub-cap [1]
Is there a preferred design for how to model empty requests (no parameters) for the API ? 
a) Use a generic TimedRequest (like this PR currently)
b) create an empty typed Request (e.g. `public class MyThingRequest extends TimedRequest{ //empty }` )  (as @gwbrown advocates)
c) introduce no arg methods in the *Client's that under the covers uses a TimedRequest (e.g. `MyThingResponse getMyThing(){ //doit }`)
d) other ?


@colings86 [2]
Any objections if I move the start/stopILM HLRC API's to the newer format and change the name to match the @dakrone's naming suggestion ? (different PR from this one)

**Comment by gwbrown** [comment]
@jakelandis #33250 has been merged into `index-lifecycle` and backported to `index-lifecycle-6.x`, so you should be good to go for switching over to the existing `OperationMode` class.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java
Comment (jakelandis) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java]: thanks missed this one. done.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/IndexLifecycleIT.java
Comment (gwbrown) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/IndexLifecycleIT.java]: @hub-cap Could you weigh in on this?

**Review by AthenaEryma — APPROVED** [review]

**Comment by gwbrown** [comment]
LGTM, pending @hub-cap's input on specialized vs generic requests.

**Comment by hub-cap** [comment]
If we create a request and we ever want to add things to it, then it will not require a breaking contract change. It also gives us a chance to rename TimedRequest if its scope changes one day. It might even be worth making it abstract to ensure we dont use it this way. what do yall think?

**Comment by jakelandis** [comment]
@hub-cap - good point. I will add the specific the request to this PR and change TimedRequest to abstract.

**Comment by jakelandis** [comment]
@dakrone @gwbrown - mind taking another look all suggestions are implemented.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/LifecycleManagementStatusResponse.java
Comment (dakrone) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/LifecycleManagementStatusResponse.java]: Can you put the `ParseField` into a class private var and then use it in the parser (so we don't accidentally typo/change it in the future)

File: client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/LifecycleManagementStatusResponse.java
Comment (dakrone) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/LifecycleManagementStatusResponse.java]: I think this should take an `OperationMode` to push the conversion into the caller, rather than doing it in the constructor (it seems cleaner to me that way).

This is just my personal preference though, so up to you if you want to change it.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/OperationMode.java
Comment (dakrone) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/OperationMode.java]: We use `Locale.ROOT` for as many things as we can, rather than the English locale here

**Review by dakrone — APPROVED** [review]
I left 3 comments but they are very minor, otherwise LGTM

File: client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/LifecycleManagementStatusResponse.java
Comment (jakelandis) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/LifecycleManagementStatusResponse.java]: moved "operation_mode" to a static final string (not exactly the ask, but i believe same outcome)

File: client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/OperationMode.java
Comment (jakelandis) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/OperationMode.java]: done

File: client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/LifecycleManagementStatusResponse.java
Comment (jakelandis) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/LifecycleManagementStatusResponse.java]: leaving as-is

**Comment by jakelandis** [comment]
Thanks for the reviews!  will merge once CI turns green.

**Comment by jakelandis** [comment]
jenkins test this please

**Pull request merged**
